### ISSUE TITLE
About the ten-crops feature extraction

### ISSUE THREAD

**Issue by Xandra298** [yujiangpu20/PEL4VAD#10]
It's so nice that you recommand the repository [video_features](https://github.com/v-iashin/video_features) for feature extraction. I have a look at the repository, and don't see any support on ten-crops/five-crops. 
Would you please share how you extract ShanghaiTech on ten-crops with that?
Best wishes!

**Comment by LucasChenLC** [comment]
Same

**Comment by yujiangpu20** [comment]
Sry for the late reply. You can extract 10-crop i3d features by following these steps:

1. Add the following code in [models/transform](https://github.com/v-iashin/video_features/blob/master/models/transforms.py). These are used for handling 10-crop augmentation, normalization, and dimension transformation.

   ```python
   class TensorTenCrop(object):
       def __init__(self, crop_size: int) -> None:
           self.crop_size = crop_size

       def __call__(self, tensor: torch.FloatTensor) -> torch.FloatTensor:
           crops = self._generate_crops(tensor)
           flipped_crops = [crop.flip(-1) for crop in crops]
           return torch.cat(crops + flipped_crops, dim=0)

       def _generate_crops(self, tensor: torch.FloatTensor) -> list:
           H, W = tensor.size(-2), tensor.size(-1)
           crop_h, crop_w = self.crop_size, self.crop_size

           # Ensure that the tensor is at least as large as the crop size
           if H < crop_h or W < crop_w:
               # Resize the tensor to be larger than the crop size, maintaining aspect ratio
               scale_factor = max(crop_h / H, crop_w / W)
               new_h, new_w = int(H * scale_factor), int(W * scale_factor)
               tensor = torch.nn.functional.interpolate(tensor.unsqueeze(0), size=(new_h, new_w), mode='bilinear', align_corners=False).squeeze(0)
               H, W = new_h, new_w

           # Coordinates for corner and center crops
           coordinates = [
               (0, 0, crop_h, crop_w),  # Top-left
               (H - crop_h, 0, H, crop_w),  # Bottom-left
               (0, W - crop_w, crop_h, W),  # Top-right
               (H - crop_h, W - crop_w, H, W),  # Bottom-right
               ((H - crop_h) // 2, (W - crop_w) // 2, (H + crop_h) // 2, (W + crop_w) // 2)  # Center
           ]

           return [self._crop_tensor(tensor, *coord) for coord in coordinates]

       def _crop_tensor(self, tensor: torch.FloatTensor, y1: int, x1: int, y2: int, x2: int) -> torch.FloatTensor:
           return tensor[..., y1:y2, x1:x2].unsqueeze(0)

   class ScaleTo1_1(object):
       def __call__(self, tensor: torch.FloatTensor) -> torch.FloatTensor:
           # Apply scaling to each crop individually
           scaled_tensors = [(2 * crop / 255) - 1 for crop in tensor]
           return torch.stack(scaled_tensors)

   class PermuteAndUnsqueeze(object):
       def __call__(self, tensor: torch.FloatTensor) -> torch.FloatTensor:
           # Apply permute and unsqueeze to each crop individually
           processed_tensors = [crop.permute(1, 0, 2, 3).unsqueeze(0) for crop in tensor]
           return torch.cat(processed_tensors, dim=0)

2. Please check the original preprocessing code located in [extract_i3d.py#L47-L60](https://github.com/v-iashin/video_features/blob/master/models/i3d/extract_i3d.py#L47-L60). Replace it with `TensorTenCrop()` and the modified `ScaleTo1_1()` and `PermuteAndUnsqueeze()`;
3. Customize your video path, which can be a text document;
4. Run the code using the following script, modify it as needed (see [video feature documentation](https://v-iashin.github.io/video_features/models/i3d/) for more details):

```python
python main.py \
    feature_type=i3d \
    streams=rgb \
    device="cuda:0" \
    on_extraction=save_numpy \
    file_with_video_paths=./sample/sample_video_paths.txt
```

5. The extracted features are named with `video_name_rgb.npy`, containing 10 crop features of the same video shaped as `(T*10, 1024)`. You can further decompose them into individual crop features.

Hope it helps~

**Comment by LucasChenLC** [comment]
Thank you very much for the detailed answer!😄

**Comment by LucasChenLC** [comment]
> Sry for the late reply. You can extract 10-crop i3d features by following these steps:
> 
> 1. Add the following code in [models/transform](https://github.com/v-iashin/video_features/blob/master/models/transforms.py). These are used for handling 10-crop augmentation, normalization, and dimension transformation.
>    ```python
>    class TensorTenCrop(object):
>        def __init__(self, crop_size: int) -> None:
>            self.crop_size = crop_size
>    
>        def __call__(self, tensor: torch.FloatTensor) -> torch.FloatTensor:
>            crops = self._generate_crops(tensor)
>            flipped_crops = [crop.flip(-1) for crop in crops]
>            return torch.cat(crops + flipped_crops, dim=0)
>    
>        def _generate_crops(self, tensor: torch.FloatTensor) -> list:
>            H, W = tensor.size(-2), tensor.size(-1)
>            crop_h, crop_w = self.crop_size, self.crop_size
>    
>            # Ensure that the tensor is at least as large as the crop size
>            if H < crop_h or W < crop_w:
>                # Resize the tensor to be larger than the crop size, maintaining aspect ratio
>                scale_factor = max(crop_h / H, crop_w / W)
>                new_h, new_w = int(H * scale_factor), int(W * scale_factor)
>                tensor = torch.nn.functional.interpolate(tensor.unsqueeze(0), size=(new_h, new_w), mode='bilinear', align_corners=False).squeeze(0)
>                H, W = new_h, new_w
>    
>            # Coordinates for corner and center crops
>            coordinates = [
>                (0, 0, crop_h, crop_w),  # Top-left
>                (H - crop_h, 0, H, crop_w),  # Bottom-left
>                (0, W - crop_w, crop_h, W),  # Top-right
>                (H - crop_h, W - crop_w, H, W),  # Bottom-right
>                ((H - crop_h) // 2, (W - crop_w) // 2, (H + crop_h) // 2, (W + crop_w) // 2)  # Center
>            ]
>    
>            return [self._crop_tensor(tensor, *coord) for coord in coordinates]
>    
>        def _crop_tensor(self, tensor: torch.FloatTensor, y1: int, x1: int, y2: int, x2: int) -> torch.FloatTensor:
>            return tensor[..., y1:y2, x1:x2].unsqueeze(0)
>    
>    class ScaleTo1_1(object):
>        def __call__(self, tensor: torch.FloatTensor) -> torch.FloatTensor:
>            # Apply scaling to each crop individually
>            scaled_tensors = [(2 * crop / 255) - 1 for crop in tensor]
>            return torch.stack(scaled_tensors)
>    
>    class PermuteAndUnsqueeze(object):
>        def __call__(self, tensor: torch.FloatTensor) -> torch.FloatTensor:
>            # Apply permute and unsqueeze to each crop individually
>            processed_tensors = [crop.permute(1, 0, 2, 3).unsqueeze(0) for crop in tensor]
>            return torch.cat(processed_tensors, dim=0)
>    ```
> 2. Please check the original preprocessing code located in [extract_i3d.py#L47-L60](https://github.com/v-iashin/video_features/blob/master/models/i3d/extract_i3d.py#L47-L60). Replace it with `TensorTenCrop()` and the modified `ScaleTo1_1()` and `PermuteAndUnsqueeze()`;
> 3. Customize your video path, which can be a text document;
> 4. Run the code using the following script, modify it as needed (see [video feature documentation](https://v-iashin.github.io/video_features/models/i3d/) for more details):
> 
> ```python
> python main.py \
>     feature_type=i3d \
>     streams=rgb \
>     device="cuda:0" \
>     on_extraction=save_numpy \
>     file_with_video_paths=your_own_data_path.txt
> ```
> 
>  5. The extracted features are named with `video_name_rgb.npy`, containing 10 crop features of the same video shaped as `(T*10, 1024)`. You can further decompose them into individual crop features.
> 
> Hope it helps~

Sorry, just one more question. What is the crop_size you use for each dataset?

**Comment by LucasChenLC** [comment]
Sorry, just one more question. What is the crop_size you use for each dataset?

**Comment by yujiangpu20** [comment]
They are set to (224, 224).

**Comment by LucasChenLC** [comment]
Thank you for the reply! Problem solved.

**Comment by modadundun** [comment]
> Thank you very much for the detailed answer!😄

Hello, how do you ensure that the extracted data's time series length matches that of the original source?Are there settings for 'extraction_fps', 'stack_size', and 'step_size'?

**Comment by pnnna123** [comment]
@modadundun Hi, I have the same question. Have you figured it out? How should they be set?
